File: test-programs/car.m
class Engine {
    def void start () {}
    def void stop () {}
}

class Driver {}

class Car {
    fd Engine engine;      // representation
    fd Driver driver;      // not representation

    Car() {
        engine = new rep Engine();
        driver = null;
    }

    def rep Engine getEngine() {
        // return engine;
    }

    void setEngine(rep Engine e) {
        engine = e;
    }

    void go() {
        // if (driver != null) engine.start();
    }
}

class Main {
    void main() {
        norep Driver bob = new norep Driver();
        norep Car car = new norep Car();
        rep Engine e;

        car.driver = bob;
        car.go();
        // car.engine.stop();           // fails
        // car.getEngine().stop();      // fails

        e = new rep Engine();
        // car.setEngine(e);            // fails
    }
}

File: src/Test/Car.hs
module Test.Car where

import AstTypes

car :: Prog
car = Prog defs [] expr

defs = [engineClass, carClass, driverClass, mainClass]

engineClass = Defn "Engine" [] [] [startM, stopM]
    where startM = Method engineType "start" [] [] $ New engineType
          stopM  = Method engineType "stop" [] [] $ New engineType

carClass = Defn "Car" [] [] [goM]

carType = OwnershipType "Car" NoRep []

goM = error ""

engineType = OwnershipType "Engine" NoRep []

driverClass = Defn "Driver" [] [] []

void = OwnershipType "void" NoRep []

mainClass = Defn "Main" [] [] [main]
    where bob     = VarDec (OwnershipType "Driver" NoRep []) "bob"
          car     = VarDec (OwnershipType "Car" NoRep []) "car"
          e       = VarDec (OwnershipType "Engine" Rep []) "e"
          carExpr = VarExpr $ VarName "car"
          main    = Method void "main" [] [bob, car, e] $ Seq [
            Asgn (VarName "bob") (New $ OwnershipType "Driver" NoRep []),
            Asgn (VarName "car") (New $ OwnershipType "Car" NoRep []),
            FieldWrite carExpr "driver" (VarExpr $ VarName "bob"),
            Invoc carExpr "go" [],
--            Invoc (FieldRead carExpr "engine") "stop" [],
--            Invoc (Invoc carExpr "getEngine" []) "stop" [],
            Asgn (VarName "e") $ New (OwnershipType "Engine" Rep [])
--            Invoc carExpr "setEngine" [VarExpr $ VarName "e"]
            ]

expr = Invoc (New $ OwnershipType "Main" NoRep []) "main" []